### ISSUE TITLE
Problem with source installation directory resulting in runtime crash

### ISSUE THREAD

**Issue by bubbleguuum** [atlas-engineer/nyxt#2953]
Current nyxt-git built for openSUSE TW with my package recipe:

https://build.opensuse.org/package/view_file/home:bobbie424242/nyxt-git/nyxt.spec?expand=1

Opening source code in the internal editor ("Open in Built-In Editor" choice in the combo box) crashes with stack trace below.

It crashes on attempting to `CHDIR` on directory /usr/share/source/mode/", a folder that does not exist. There are 2 problems:

- that it crashes because of a missing folder instead of failing gracefully
- that it looks for the wrong installed source dir (`/usr/share/source/`) instead of `/usr/share/nyxt/source` where it is actually installed.

I know this worked at some point. Is there something that must be changed in flags passed to `make` (see .spec file linked above)  or is it a nyxt build bug ?

////

```
Unhandled SB-POSIX:SYSCALL-ERROR in thread #<SB-THREAD:THREAD tid=7140 "Anonymous thread" RUNNING
                                              {100B2DE9F3}>:
  Error in SB-POSIX:CHDIR: No such file or directory (2)

Backtrace for: #<SB-THREAD:THREAD tid=7140 "Anonymous thread" RUNNING {100B2DE9F3}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-POSIX:SYSCALL-ERROR {100B310A93}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK SB-EXT:*INVOKE-DEBUGGER-HOOK* #<SB-POSIX:SYSCALL-ERROR {100B310A93}>)
2: (INVOKE-DEBUGGER #<SB-POSIX:SYSCALL-ERROR {100B310A93}>)
3: (ERROR SB-POSIX:SYSCALL-ERROR :NAME SB-POSIX:CHDIR :ERRNO 2)
4: (SB-POSIX:SYSCALL-ERROR SB-POSIX:CHDIR)
5: (SB-POSIX:CHDIR "/usr/share/source/mode/")
6: ((LAMBDA (URL BUFFER) :IN "/home/abuild/rpmbuild/BUILD/nyxt-3prerelease6+git20230509.f45330f6/source/mode/editor.lisp") "editor:/usr/share/source/mode/bookmark.lisp" #<NYXT/MODE/EDITOR:EDITOR-BUFFER 1904 {1008C53303}>)
7: ((LAMBDA (NYXT/RENDERER/GTK::REQUEST) :IN NYXT/RENDERER/GTK::MAKE-CONTEXT) #<CL-WEBKIT2:WEBKIT-URI-SCHEME-REQUEST {100B2E12A3}>)
8: ((LAMBDA NIL :IN CL-WEBKIT2::URI-SCHEME-PROCESSED))
9: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
10: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
11: ((FLET "WITHOUT-INTERRUPTS-BODY-156" :IN SB-THREAD::RUN))
12: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
13: ((FLET "WITHOUT-INTERRUPTS-BODY-149" :IN SB-THREAD::RUN))
14: (SB-THREAD::RUN)
15: ("foreign function: call_into_lisp_")
```

**Comment by aadcg** [comment]
> I know this worked at some point.

The change that may be responsible is commit e296a2e83cc67f1c181f93641c8d3619ff35bc0b.

**Comment by Ambrevar** [comment]
Possibly 3 problems here:

- The source location resolution seems wrong in this case.
- We should handle non-existing files gracefully.
- Why does it crash?  Looks like we are missing a `with-protect` somewhere.

**Comment by aadcg** [comment]
Concerning the source location, the relevant variables are the ones below. @bubbleguuum, I believe you should tweak env vars `PREFIX` and `DATADIR` to suits your needs. Is there something we should do from our side?

```lisp
(export-always '*prefix*)
(defparameter *prefix* (merge-pathnames* (relative-path-from-env "PREFIX" #p"usr/local/")
                                         *destdir*))

(export-always '*datadir*)
(defparameter *datadir* (path-from-env "DATADIR" (merge-pathnames* "share/" *prefix*)))
```

**Comment by bubbleguuum** [comment]
Invoking make with  `DATADIR=/usr/share/nyxt` did indeed fix the source path.
This was not necessary previously. So if you do not set it, `DATADIR` seems to default to `/usr/share` (or rather `$PREFIX/share`). Maybe it should default to `$PREFIX/share/nyxt` ?

 I found another crash: if you have a function whose code is opened in Nyxt and in the combo box (now on the right of the window) you reselect "Open in Built-In Editor" (which admittedly does not make much sense) , it crashes with this stack trace:

```
Unhandled SB-PCL::NO-APPLICABLE-METHOD-ERROR in thread #<SB-THREAD:THREAD tid=1171 "Anonymous thread" RUNNING
                                                          {100C4E9263}>:
  There is no applicable method for the generic function
    #<STANDARD-GENERIC-FUNCTION NYXT/MODE/EDITOR:MARKUP (2)>
  when called with arguments
    (NIL).
See also:
  The ANSI Standard, Section 7.6.6

Backtrace for: #<SB-THREAD:THREAD tid=1171 "Anonymous thread" RUNNING {100C4E9263}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-PCL::NO-APPLICABLE-METHOD-ERROR {100C508C63}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK SB-EXT:*INVOKE-DEBUGGER-HOOK* #<SB-PCL::NO-APPLICABLE-METHOD-ERROR {100C508C63}>)
2: (INVOKE-DEBUGGER #<SB-PCL::NO-APPLICABLE-METHOD-ERROR {100C508C63}>)
3: (ERROR SB-PCL::NO-APPLICABLE-METHOD-ERROR :GENERIC-FUNCTION #<STANDARD-GENERIC-FUNCTION NYXT/MODE/EDITOR:MARKUP (2)> :ARGS (NIL))
4: ((:METHOD NO-APPLICABLE-METHOD (T)) #<STANDARD-GENERIC-FUNCTION NYXT/MODE/EDITOR:MARKUP (2)> NIL) [fast-method]
5: (SB-PCL::CALL-NO-APPLICABLE-METHOD #<STANDARD-GENERIC-FUNCTION NYXT/MODE/EDITOR:MARKUP (2)> (NIL))
6: ((LAMBDA (NYXT/RENDERER/GTK::REQUEST) :IN NYXT/RENDERER/GTK::MAKE-CONTEXT) #<CL-WEBKIT2:WEBKIT-URI-SCHEME-REQUEST {100C3BF7C3}>)
7: ((LAMBDA NIL :IN CL-WEBKIT2::URI-SCHEME-PROCESSED))
8: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
9: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
10: ((FLET "WITHOUT-INTERRUPTS-BODY-156" :IN SB-THREAD::RUN))
11: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
12: ((FLET "WITHOUT-INTERRUPTS-BODY-149" :IN SB-THREAD::RUN))
13: (SB-THREAD::RUN)
14: ("foreign function: call_into_lisp_")

```

**Comment by Ambrevar** [comment]
@aartaka Can you have a look at the crashes? (Or maybe @aadcg if you are familiar with this code.)
I'll have a look at the DATADIR issue.

**Comment by bubbleguuum** [comment]
For completeness, there is a discrepancy (with `DATADIR` not set)  between `make install` installing source in `$PREFIX/share/nyxt` and runtime looking for it in `$PREFIX/share`

**Comment by aadcg** [comment]
> I found another crash: if you have a function whose code is opened in Nyxt and in the combo box (now on the right of the window) you reselect "Open in Built-In Editor" (which admittedly does not make much sense) , it crashes with this stack trace:

Could you explain in more detail what doesn't make sense? 

Just to make sure we're on the same page: when you say "have a function whose code is opened in Nyxt", that meansthat you've issued command `describe-funtion`, correct? 

My understanding is that an editor buffer is being created and `editor-mode` isn't enabled. But I'm not particularly familiar with this part of the codebase and I've used these features. Maybe @aartaka can help.

**Comment by Ambrevar** [comment]
@bubbleguuum Can you report the value that `asdf:*default-source-registries*` has in Nyxt?

**Comment by Ambrevar** [comment]
Also can you share what `(asdf:system-source-directory :nyxt)` returns ?

**Comment by bubbleguuum** [comment]
That's with nyxt compiled with `PREFIX=/usr` and `DATADIR=/usr/share/nyxt`:

` asdf:*default-source-registries*` returns `(NYXT-SOURCE-REGISTRY ENVIRONMENT-SOURCE-REGISTRY)`. 

`(asdf:system-source-directory :nyxt)` returns `/home/abuild/rpmbuild/BUILD/nyxt-3prerelease6+git20230509.f45330f6/` which is the path used for the build and should not be used at runtime (apparently, it isn't).

On launch, nyxt displays the proper source path (again with the proper DATADIR set):

```
<INFO> [12:13:07] Source location: #P"/usr/share/nyxt/"
```

Without DATADIR, it would display '/usr/share' above

**Comment by Ambrevar** [comment]
Oh, right, that makes sense, thanks for sharing.

**Comment by bubbleguuum** [comment]
> 


> Could you explain in more detail what doesn't make sense?
> 
> Just to make sure we're on the same page: when you say "have a function whose code is opened in Nyxt", that means that you've issued command `describe-funtion`, correct?

I attached a screenshot to make it clearer what triggers the crash 
The "does not make much sense mention" was about  selecting "Open in Built-In Editor" in that menu while nyxt is already in that mode (if I understand this correctly).



![nyxt](https://github.com/atlas-engineer/nyxt/assets/761164/b520cad5-0055-4e70-9bc5-e28f937481b9)

**Comment by aadcg** [comment]
@bubbleguuum I understand your confusion. You'd expect that you'd be able to edit the command from that same buffer. 

Clicking on "Open in Built-In Editor" should open a new buffer for the sole purpose of editing.

**Comment by bubbleguuum** [comment]
Ah yes indeed was confused about this !

**Comment by aadcg** [comment]
The editor bug is fixed by e856eb878b1a13487a97a201027d24349b8339e4.

With respect to the source directory, #2955 fixes it.

**Comment by Ambrevar** [comment]
Wait, @bubbleguuum can you confirm that #2955 fixes the source dir?

**Comment by bubbleguuum** [comment]
Current git with #2955 applied does not compile:

```
[   94s] ; compiling file "NYXT:SOURCE;CONFIGURATION.LISP.NEWEST" (written 11 MAY 2023 08:36:51 AM):
[   94s] ; 
[   94s] ; caught ERROR:
[   94s] ;   READ error during COMPILE-FILE:
[   94s] ;   
[   94s] ;     The symbol "*DEST-SOURCE-FINAL-DIR*" is not external in the NASDF package.
[   94s] ;   
[   94s] ;       Line: 103, Column: 37, File-Position: 4195
[   94s] ;   
[   94s] ;       Stream: #<SB-INT:FORM-TRACKING-STREAM for "file /home/abuild/rpmbuild/BUILD/nyxt-3prerelease6+git20230511.20f83ed/source/configuration.lisp" {1009AA7E63}>
[   94s] 
[   94s] ; compilation aborted after 0:00:00.028
[   94s] Unhandled UIOP/LISP-BUILD:COMPILE-FILE-ERROR in thread #<SB-THREAD:THREAD tid=8140 "main thread" RUNNING
[   94s]                                                           {10013B0113}>:
[   94s]   COMPILE-FILE-ERROR while
[   94s]   compiling #<CL-SOURCE-FILE "nyxt" "Core" "configuration">
[   94s] 
[   94s] Backtrace for: #<SB-THREAD:THREAD tid=8140 "main thread" RUNNING {10013B0113}>
[   94s] 0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<UIOP/LISP-BUILD:COMPILE-FILE-ERROR {100ACFC543}> #<unused argument> :QUIT T)
[   94s] 1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<UIOP/LISP-BUILD:COMPILE-FILE-ERROR {100ACFC543}>)
[   94s] 2: (INVOKE-DEBUGGER #<UIOP/LISP-BUILD:COMPILE-FILE-ERROR {100ACFC543}>)
[   94s] 3: (ERROR UIOP/LISP-BUILD:COMPILE-FILE-ERROR :CONTEXT-FORMAT "~/asdf-action::format-action/" :CONTEXT-ARGUMENTS ((#<ASDF/LISP-ACTION:COMPILE-OP > . #<ASDF/LISP-ACTION:CL-SOURCE-FILE "nyxt" "Core" "configuration">)))
[   94s] 4: (UIOP/LISP-BUILD:CHECK-LISP-COMPILE-RESULTS NIL T T "~/asdf-action::format-action/" ((#<ASDF/LISP-ACTION:COMPILE-OP > . #<ASDF/LISP-ACTION:CL-SOURCE-FILE "nyxt" "Core" "configuration">)))
[   94s] 5: ((SB-PCL::EMF ASDF/ACTION:PERFORM) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/LISP-ACTION:CL-SOURCE-FILE "nyxt" "Core" "configuration">)
[   94s] 6: ((LAMBDA NIL :IN ASDF/ACTION:CALL-WHILE-VISITING-ACTION))
[   94s] 7: ((:METHOD ASDF/ACTION:PERFORM-WITH-RESTARTS :AROUND (T T)) #<ASDF/LISP-ACTION:COMPILE-OP > #<ASDF/LISP-ACTION:CL-SOURCE-FILE "nyxt" "Core" "configuration">) [fast-method]
[   94s] 8: ((:METHOD ASDF/PLAN:PERFORM-PLAN (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1004673343}>) [fast-method]
[   94s] 9: ((FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT))
[   94s] 10: ((:METHOD ASDF/PLAN:PERFORM-PLAN :AROUND (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1004673343}>) [fast-method]
[   94s] 11: ((:METHOD ASDF/OPERATE:OPERATE (ASDF/OPERATION:OPERATION ASDF/COMPONENT:COMPONENT)) #<ASDF/LISP-ACTION:LOAD-OP > #<NASDF:NASDF-SYSTEM "nyxt/gi-gtk-application"> :PLAN-CLASS NIL :PLAN-OPTIONS NIL) [fast-method]

```

**Comment by Ambrevar** [comment]
That's because NASDF is found in Nclasses & friends first, thus it does not know the new symbol.

We'd need to deploy this change to other libraries first. Steps

- Update NASDF in Ntemplate, Nclasses, etc.
- Update all the Nclasses & friends submodules in Nyxt.
- Update all the Guix package.

But first:
- I'll test this more thoroughly so that we don't have to iterate too many times.
- @aartaka Shall we merge #2925 first ?

@aadcg @aartaka This procedure of updating NASDF everywhere is cumbersome.  Any idea how we could do better?

**Comment by aadcg** [comment]
@Ambrevar sorry for closing the issue before it was time.

I think merging #2955 is more important than #2925. It would be better to update NASDF once, rather than twice nonetheless (i.e. rebase #2955 on top of #2925). I think you should decide about it with @aartaka.

With respect to this cumbersome process, I'm afraid we can't do much about it.

**Comment by aartaka** [comment]
> I think merging #2955 is more important than #2925. It would be better to update NASDF once, rather than twice nonetheless (i.e. rebase #2955 on top of #2925). I think you should decide it with @aartaka.

I'm okay with NASDF merged later, yes.

> With respect to this cumbersome process, I'm afraid we can't do much about it.

Maybe we can use a better templating library os some ASDF-resident code generation? ( ͡° ͜ʖ ͡°)

Not sure how to even approach this, though.

**Comment by Ambrevar** [comment]
See #2955, we should not have to update NASDF anywhere.

**Comment by Ambrevar** [comment]
Source directory resolution is fixed!  Now the editor-mode code remains to be fixed...

**Comment by aadcg** [comment]
editor-mode has been fixed in e856eb878b1a13487a97a201027d24349b8339e4.
